Proactively cleanup nw_resolve requests in case they don't progress.

nmahendru opened the pull request:
#### 163949dce4720d4affa6cdcc4e7c28811b580e32
<pre>
Proactively cleanup nw_resolve requests in case they don&apos;t progress.
<a href="https://bugs.webkit.org/show_bug.cgi?id=279997">https://bugs.webkit.org/show_bug.cgi?id=279997</a>
<a href="https://rdar.apple.com/133085594">rdar://133085594</a>

Reviewed by Chris Dumez and Youenn Fablet.

We have noticed huge number of pending requests in some cases. They are
not progressing for some reason and this change is to do a proactive cleanup.
60 seconds should be enough to resolve or cancel.

It is possible that a change is also needed on the networking stack.

* Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
(WebCore::DNSResolveQueueCFNet::performDNSLookup):

Canonical link: <a href="https://commits.webkit.org/284012@main">https://commits.webkit.org/284012@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/71ebb5005b71d32c21959e7898efb8358ca26e26

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68063 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47455 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20722 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72126 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19208 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70180 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55251 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19024 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54399 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/12812 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71130 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43452 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58821 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34861 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40121 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16237 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17565 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62092 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16578 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73822 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12034 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15857 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/61851 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12072 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58896 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/61867 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9772 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3391 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10366 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43256 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44329 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45524 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44071 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
```diff
@@ -117,10 +117,20 @@ void DNSResolveQueueCFNet::performDNSLookup(const String& hostname, Ref<Completi
 
-    nw_resolver_set_update_handler(resolver.get(), dispatch_get_main_queue(), makeBlockPtr([resolver, completionHandler = WTFMove(completionHandler)] (nw_resolver_status_t status, nw_array_t resolvedEndpoints) mutable {
+    auto timer = makeUnique<Timer>([resolver, completionHandler]() {
+        WTF::ensureOnMainThread([resolver, completionHandler]() {
```
You don't need `WTF::`

cdumez replied:
We can use `resolver = WTFMove(resolver), completionHandler = WTFMove(completionHandler)` here to avoid refcounting churn.

cdumez commented on Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
```diff
@@ -117,10 +117,20 @@ void DNSResolveQueueCFNet::performDNSLookup(const String& hostname, Ref<Completi
     RetainPtr resolver = adoptCF(nw_resolver_create_with_endpoint(hostEndpoint.get(), parameters.get()));
 
-    nw_resolver_set_update_handler(resolver.get(), dispatch_get_main_queue(), makeBlockPtr([resolver, completionHandler = WTFMove(completionHandler)] (nw_resolver_status_t status, nw_array_t resolvedEndpoints) mutable {
+    auto timer = makeUnique<Timer>([resolver, completionHandler]() {
```
`() mutable {` because of the change I'm suggesting below.

cdumez commented on Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
```diff
@@ -117,10 +117,20 @@ void DNSResolveQueueCFNet::performDNSLookup(const String& hostname, Ref<Completi
+            completionHandler->complete(makeUnexpected(DNSError::Cancelled));
+        });
+    });
+    timer->startOneShot(Seconds(60));
```
`Seconds(60)` -> `60_s`

youennf replied:
```suggestion
    timer->startOneShot(60_s);
```
But usually we would introduce a `static constexpr Seconds timeoutForDNSResolution = 60_s;`

cdumez commented on Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
```diff
@@ -117,10 +117,20 @@ void DNSResolveQueueCFNet::performDNSLookup(const String& hostname, Ref<Completi
-        auto callCompletionHandler = [resolver = WTFMove(resolver), completionHandler = WTFMove(completionHandler)](DNSAddressesOrError&& result) mutable {
+
+        auto callCompletionHandler = [resolver, completionHandler = WTFMove(completionHandler), timer = WTFMove(timer)](DNSAddressesOrError&& result) mutable {
+            timer->stop();
```
May be nice to rename this timer to `timeoutTimer` to make it clearer what it's for.

youennf commented on Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
```diff
@@ -117,10 +117,20 @@ void DNSResolveQueueCFNet::performDNSLookup(const String& hostname, Ref<Completi
+    timer->startOneShot(Seconds(60));
+
+    nw_resolver_set_update_handler(resolver.get(), dispatch_get_main_queue(), makeBlockPtr([resolver, completionHandler = WTFMove(completionHandler), timer = WTFMove(timer)] (nw_resolver_status_t status, nw_array_t resolvedEndpoints) mutable {
         if (status == nw_resolver_status_in_progress)
```
Let's change it to `status != nw_resolver_status_complete`

nmahendru replied:
done.

cdumez reviewed (approved):
LGTM with nits.

youennf commented on Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
```diff
@@ -117,10 +117,20 @@ void DNSResolveQueueCFNet::performDNSLookup(const String& hostname, Ref<Completi
 
-    nw_resolver_set_update_handler(resolver.get(), dispatch_get_main_queue(), makeBlockPtr([resolver, completionHandler = WTFMove(completionHandler)] (nw_resolver_status_t status, nw_array_t resolvedEndpoints) mutable {
+    auto timer = makeUnique<Timer>([resolver, completionHandler]() {
+        WTF::ensureOnMainThread([resolver, completionHandler]() {
```
Timer AFAIK needs to be created on main thread and is executing its callback on main thread.
This call to ensureOnMainThread is probably not needed.
Otherwise, we should use `resolver = WTFMove(resolver), completionHandler = WTFMove(completionHandler)` to save some refcounting (and making the outer lambda mutable.

nmahendru replied:
done.

youennf commented on Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
```diff
@@ -117,10 +117,20 @@ void DNSResolveQueueCFNet::performDNSLookup(const String& hostname, Ref<Completi
             return;
-        auto callCompletionHandler = [resolver = WTFMove(resolver), completionHandler = WTFMove(completionHandler)](DNSAddressesOrError&& result) mutable {
+
+        auto callCompletionHandler = [resolver, completionHandler = WTFMove(completionHandler), timer = WTFMove(timer)](DNSAddressesOrError&& result) mutable {
```
Why not `resolver = WTFMove(resolver)` here?

nmahendru replied:
added.

youennf commented on Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
```diff
@@ -117,10 +117,20 @@ void DNSResolveQueueCFNet::performDNSLookup(const String& hostname, Ref<Completi
+    });
+    timer->startOneShot(Seconds(60));
+
+    nw_resolver_set_update_handler(resolver.get(), dispatch_get_main_queue(), makeBlockPtr([resolver, completionHandler = WTFMove(completionHandler), timer = WTFMove(timer)] (nw_resolver_status_t status, nw_array_t resolvedEndpoints) mutable {
```
In case this got cancelled, I wonder in which thread this block is destroyed.
It should not be destroyed out of main thread, timer should be destroyed in main thread. Can you validate this?

nmahendru replied:
I think the handler is called on the main thread because of dispatch_get_main_queue() so it will always be run on the main thread and then the de-alloc will also happen on the main thread ?

youennf replied:
The callback will always be executed in main thread, but I am less sure about the deallocation in case the callback is not called for instance. Could you have a try by running that code with a timeout of 0 so that cancel happens? Timer or TimerBase probably have ASSERT/RELEASE_ASSERT in their destructors.

To trigger mdns resolution, you can probably use https://webrtc.github.io/samples/src/content/datachannel/basic/.

youennf reviewed (approved):
Look good to me overall, some comments below

cdumez commented on Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
```diff
@@ -117,10 +119,19 @@ void DNSResolveQueueCFNet::performDNSLookup(const String& hostname, Ref<Completi
+        nw_resolver_cancel(resolver.get());
+        completionHandler->complete(makeUnexpected(DNSError::Cancelled));
+    });
+    timeoutTimer->startOneShot(Seconds(timeoutForDNSResolution));
```
You don't need `Seconds()` here, `timeoutForDNSResolution` already has the right type.

cdumez reviewed (approved):
LGTM with nit.

nmahendru committed:
Proactively cleanup nw_resolve requests in case they don't progress.
https://bugs.webkit.org/show_bug.cgi?id=279997
rdar://133085594

Reviewed by Chris Dumez and Youenn Fablet.

We have noticed huge number of pending requests in some cases. They are
not progressing for some reason and this change is to do a proactive cleanup.
60 seconds should be enough to resolve or cancel.

It is possible that a change is also needed on the networking stack.

* Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
(WebCore::DNSResolveQueueCFNet::performDNSLookup):

Canonical link: https://commits.webkit.org/284012@main
